pkg/identity/cache.go

```diff
@@ -63,13 +65,44 @@ func GetIdentities() []*models.Identity {
 	return identities
 }
 
-func identityWatcher(owner IdentityAllocatorOwner) {
+func identityWatcher(owner IdentityAllocatorOwner, events allocator.AllocatorEventChan) {
+	var (
+		needPolicyUpdateMutex lock.RWMutex
+		needPolicyUpdate      bool
+	)
+
+	// The event queue handler is kept as lightweight as possible, it only
+	// sets a flag to signal the requirement to trigger policy
+	// recalculation.  A separate go routine in the background checks for
+	// the flag and sequentially invokes the policy update function. After
+	// an update period, the flag is immediately checked again. If no
+	// update was required, the routine sleeps for a second and tries again
+	go func() {
+		for {
+			triggerPolicyUpdate := false
+
+			// keep critical section as small as possible
+			needPolicyUpdateMutex.Lock()
+			triggerPolicyUpdate = needPolicyUpdate
+			needPolicyUpdate = false
+			needPolicyUpdateMutex.Unlock()
+
+			if triggerPolicyUpdate {
+				owner.TriggerPolicyUpdates(true)
+			} else {
+				time.Sleep(time.Second)
+			}
+		}
+	}()
+
 	for {
-		event := <-identityAllocator.Events
+		event := <-events
 
 		switch event.Typ {
 		case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
-			owner.TriggerPolicyUpdates(true)
+			needPolicyUpdateMutex.Lock()
+			needPolicyUpdate = true
```

Comment:
Wait, can't this be a channel? If this becomes a channel, you don't need to to a time.Sleep on line 91

Reply:
The major change here is to avoid triggering a policy for every event so aggregation has to happen. We can do a unbuffered channel with non-blocking writes though.